### PR TITLE
Kinetic backend

### PR DESCRIPTION
**Pull request by lamphamsy**
Added kinetic backend into S3 with put/get/delete operations
It support currently object's size <= 1MB
@AntoninCoulibaly : I developed this from your branch (https://github.com/scality/S3/tree/dev/ft/kineticBackend)

### CODE REVIEW

File: lib/Config.js [hunk 1]
```diff
@@ -263,6 +263,13 @@ class Config {
             dataPath,
             metadataPath,
         };
+
+        if (config.kinetic && Array.isArray(config.kinetic)
+           && process.env.S3DATA === 'kinetic') {
+            this.kinetic = {};
+            this.kinetic.host = config.kinetic[0].split(':')[0];
```
Comment (ThibaultRiviere) [inline lib/Config.js]: Should we trigger an error if the config kinetic doesn't exist ?

File: lib/data/kinetic/backend.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+export const ds = [];
+
+const kDrives = {
+    port: config.kinetic ? config.kinetic.port : 8123,
+    host: config.kinetic ? config.kinetic.host : 'localhost',
+};
+
+function newSocket() {
+    const socket = new net.Socket({ allowHalfOpen: false });
```
Comment (ThibaultRiviere) [inline lib/data/kinetic/backend.js]: why the ```allowHalfOpen```?

File: lib/data/kinetic/backend.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        // This message contains only status of response. Process 2nd message
+        return decodeRecMsg(socket, cb);
+    });
+}
+
+const backend = {
+    put: function x(request, size, keyContext, reqUids, callback) {
+        const value = [];
+        request.on('data', data => {
+            value.push(data);
```
Comment (ThibaultRiviere) [inline lib/data/kinetic/backend.js]: If you wait the event ```end``` that mean you keep all the data in memory?

File: lib/Config.js [hunk 4]
```diff
@@ -234,8 +234,8 @@ class Config {
             }
             this.authData = authData;
         }
-        if (process.env.S3SPROXYD) {
-            data = process.env.S3SPROXYD;
+        if (process.env.S3DATA) {
+            data = process.env.S3DATA;
```
Comment (AntoninCoulibaly) [inline lib/Config.js]: Can you put this change in another commit ?

File: lib/data/kinetic/backend.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
+import crypto from 'crypto';
+import kinetic from 'kineticlib';
+import stream from 'stream';
+import net from 'net';
+
+import { errors } from 'arsenal';
+
+import config from '../../Config';
```
Comment (AntoninCoulibaly) [inline lib/data/kinetic/backend.js]: Is it possible to have some logs ?

File: lib/data/kinetic/backend.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+let sequence = 1;
+
+export const ds = [];
```
Comment (AntoninCoulibaly) [inline lib/data/kinetic/backend.js]: It is not used right ?

File: lib/data/kinetic/backend.js [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        socket.write(pdu.read(), err => {
+            if (err) {
+                return callback(err);
+            }
+            let statusCode;
+            socket.on('data', chunk => {
+                const recPDU = new kinetic.PDU(chunk);
+                statusCode = recPDU.getStatusCode();
+                socket.end();
+            }).on('end', err => {
```
Comment (AntoninCoulibaly) [inline lib/data/kinetic/backend.js]: I think it could be great to have smaller functions

File: lib/Config.js [hunk 8]
```diff
@@ -263,6 +263,13 @@ class Config {
             dataPath,
             metadataPath,
         };
+
+        if (config.kinetic && Array.isArray(config.kinetic)
+           && process.env.S3DATA === 'kinetic') {
```
Comment (GuillaumeGomez) [inline lib/Config.js]: The "&&" should be preceded by one more space.

File: lib/data/kinetic/backend.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    port: config.kinetic ? config.kinetic.port : 8123,
+    host: config.kinetic ? config.kinetic.host : 'localhost',
+};
+
+function propError(pdu) {
+    const statusCode = pdu.getStatusCode();
+    let err;
+
+    if (statusCode !== kinetic.errors.SUCCESS) {
+        if (statusCode === kinetic.errors.NOT_FOUND) {
```
Comment (GuillaumeGomez) [inline lib/data/kinetic/backend.js]: I would have wrote this as follow:

```js
        if (statusCode === kinetic.errors.NOT_FOUND) {
            return errors.ObjNotFound;
        }
        return kinetic.getErrorName(statusCode);
```

And so the `err` variable becomes useless.

File: lib/data/kinetic/backend.js [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            err = errors.ObjNotFound;
+        } else {
+            err = kinetic.getErrorName(statusCode);
+        }
+        return err;
+    }
+}
```
Comment (GuillaumeGomez) [inline lib/data/kinetic/backend.js]: I think a return statement is missing here.

File: lib/data/kinetic/backend.js [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    const pdu = new kinetic.PutPDU(sequence, key, obj.length, tag, options);
+    ++sequence;
+
+    socket.write(Buffer.concat([pdu.read(), obj]), err => {
+        if (err) {
+            return callback(err);
+        }
+        kinetic.streamToPDU(socket, (err, pdu) => {
+            const err1 = propError(pdu);
+            socket.destroy();
```
Comment (lamphamsy) [inline lib/data/kinetic/backend.js]: Should it be better with `socket.end()` instead of destroying the socket?

File: lib/data/kinetic/backend.js [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+function getKinetic(socket, pdu, callback) {
+    let count = 0;
+    const value = [];
+    const err = propError(pdu);
+    socket.resume();
+    socket.on('data', chunk => {
+        value.push(chunk);
+        count += chunk.length;
+        if (count === pdu.getChunkSize()) {
```
Comment (lamphamsy) [inline lib/data/kinetic/backend.js]: object size should be calculated once before here

File: lib/data/kinetic/backend.js [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    socket.write(pdu.read(), err => {
+        if (err) {
+            return callback(err);
+        }
+        kinetic.streamToPDU(socket, (err, pdu) => {
+            if (err) {
+                return callback(err);
+            }
+            const err1 = propError(pdu);
+            socket.destroy();
```
Comment (lamphamsy) [inline lib/data/kinetic/backend.js]: same here: `socket.end()`?

File: lib/data/kinetic/backend.js [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    const socket = new net.Socket({ allowHalfOpen: false }).pause();
+    socket.setKeepAlive(true);
+    socket.unref();
+    return socket.connect(kDrives.port, kDrives.host, () => {
+        kinetic.streamToPDU(socket, (err, pdu) => {
+            if (err) {
+                return callback(err);
+            }
+            const err1 = propError(pdu);
+            return callback(err1, socket, pdu.getChunkSize());
```
Comment (lamphamsy) [inline lib/data/kinetic/backend.js]: 3rd arg is not used

File: lib/Config.js [hunk 15]
```diff
@@ -263,6 +263,13 @@ class Config {
             dataPath,
             metadataPath,
         };
+
+        if (config.kinetic && Array.isArray(config.kinetic)
+            && process.env.S3DATA === 'kinetic') {
+            this.kinetic = {};
```
Comment (MathieuCassagne) [inline lib/Config.js]: You can write it like that :)
```
const tab = config.kinetic[0].split(':');
this.kinetic = {
    host: tab[0],
    port: tab[1],
};
```
You avoid one split, and you do not add property to the object.

File: lib/data/kinetic/backend.js [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+import net from 'net';
+
+import { errors } from 'arsenal';
+
+import config from '../../Config';
+
+let sequence = 1;
+
+const kDrives = {
+    port: config.kinetic ? config.kinetic.port : 8123,
```
Comment (MathieuCassagne) [inline lib/data/kinetic/backend.js]: You do the ternaire twice, maybe you can do it only once.

File: lib/data/kinetic/backend.js [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+        return pdu.getErrorMessage();
+    }
+    return undefined;
+}
+
+function newSocket(callback) {
+    const socket = new net.Socket({ allowHalfOpen: false }).pause();
+    socket.setKeepAlive(true);
+    socket.unref();
+    return socket.connect(kDrives.port, kDrives.host, () => {
```
Comment (MathieuCassagne) [inline lib/data/kinetic/backend.js]: Here you can just pass the kDrives object to connect as first argument.
https://nodejs.org/dist/latest-v4.x/docs/api/net.html#net_socket_connect_options_connectlistener

File: lib/data/kinetic/backend.js [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+function putKinetic(socket, value, options, callback) {
+    const key = crypto.randomBytes(20);
+    const obj = Buffer.concat(value);
+    const tag = crypto.createHmac("sha1", "asdfasdf").update(obj).digest();
+    const pdu = new kinetic.PutPDU(sequence, key, obj.length, tag, options);
+    ++sequence;
+
+    socket.write(Buffer.concat([pdu.read(), obj]), err => {
```
Comment (MathieuCassagne) [inline lib/data/kinetic/backend.js]: Can you know the size of ```pdu.read()``` in advance, and pass the resultin length to Buffer.concat ? The doc says that it will be faster if you do so, it'll require one less loop.
https://nodejs.org/dist/latest-v4.x/docs/api/buffer.html#buffer_class_method_buffer_concat_list_totallength

File: lib/data/kinetic/backend.js [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+        return undefined;
+    });
+}
+
+function getKinetic(socket, pdu, callback) {
+    let count = 0;
+    const chunkSize = pdu.getChunkSize();
+    const value = [];
+    const err = propError(pdu);
+    socket.resume();
```
Comment (MathieuCassagne) [inline lib/data/kinetic/backend.js]: You do not need to resume the socket here, since you put a ```socket.on('data')``` the socket will pass in the flowing mode automatically.
https://nodejs.org/dist/latest-v4.x/docs/api/stream.html#stream_class_stream_readable

Comment (AntoninCoulibaly) [inline lib/data/kinetic/backend.js]: The socket need to be resume because it is intentionally paused before, so we need to resume it for getting data.

**Comment by AntoninCoulibaly** [comment]
@GuillaumeGomez @MathieuCassagne  comments addressed!

File: lib/data/kinetic/backend.js (hunk 19 above)
Comment (MathieuCassagne) [inline lib/data/kinetic/backend.js]: I might misunderstood the docs, nevermind :)

File: lib/Config.js [hunk 20]
```diff
@@ -234,8 +234,8 @@ class Config {
             }
             this.authData = authData;
         }
-        if (process.env.S3SPROXYD) {
-            data = process.env.S3SPROXYD;
+        if (process.env.S3DATA) {
```
Comment (michael-zapata) [inline lib/Config.js]: Careful with that change. I think our internal process uses `S3SPROXYD` so you will need to update there before merging here.

File: lib/data/kinetic/backend.js [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+        });
+    });
+}
+
+function putKinetic(socket, value, options, callback) {
+    const key = crypto.randomBytes(20);
+    const obj = Buffer.concat(value);
+    const tag = crypto.createHmac('sha1', 'asdfasdf').update(obj).digest();
+    const pdu = new kinetic.PutPDU(sequence, key, obj.length, tag, options);
+    ++sequence;
```
Comment (michael-zapata) [inline lib/data/kinetic/backend.js]: This will overflow at some point, do we really want that? (That goes for all occurrences)

File: lib/data/kinetic/backend.js [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+    const header = pdu.read();
+    const len = header.length + obj.length;
+
+    socket.write(Buffer.concat([header, obj], len), err => {
+        if (err) {
+            return callback(err);
+        }
+        kinetic.streamToPDU(socket, (err, pdu) => {
+            const err1 = propError(pdu);
+            socket.destroy();
```
Comment (michael-zapata) [inline lib/data/kinetic/backend.js]: Performance concern: Can't we keep sockets alive instead of destroying them systematically?

File: lib/data/kinetic/backend.js [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+        });
+        return undefined;
+    });
+}
+
+const backend = {
+    put: function putK(request, size, keyContext, reqUids, callback) {
+        const value = [];
+        request.on('data', data => {
+            value.push(data);
```
Comment (michael-zapata) [inline lib/data/kinetic/backend.js]: Are we really putting all the data in memory before writing it into the socket?

File: lib/data/kinetic/backend.js [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+    socket.write(Buffer.concat([header, obj], len), err => {
+        if (err) {
+            return callback(err);
+        }
+        kinetic.streamToPDU(socket, (err, pdu) => {
+            const err1 = propError(pdu);
+            socket.destroy();
+            return callback(err || err1, key);
+        });
+        return undefined;
```
Comment (michael-zapata) [inline lib/data/kinetic/backend.js]: Instead of returning `undefined`, you can return `kinetic.streamToPDU()` instead, that'll avoid a few lines in every function.

File: lib/data/kinetic/backend.js [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+    });
+}
+
+function getKinetic(socket, pdu, callback) {
```
Comment (michael-zapata) [inline lib/data/kinetic/backend.js]: This looks like we are putting everything in memory before even trying to send some data back. Could we avoid that?

File: lib/data/kinetic/backend.js (hunk 23 above)
Comment (AntoninCoulibaly) [inline lib/data/kinetic/backend.js]: Yes that is what we wrote in the commit message that for now we are not handling more than 1mB files we are currently on it.
Our idea was to split into multiple 1MB files.

File: lib/Config.js (hunk 20 above)
Comment (AntoninCoulibaly) [inline lib/Config.js]: I will remove this change to do it in a different PR.

File: lib/data/kinetic/backend.js (hunk 22 above)
Comment (AntoninCoulibaly) [inline lib/data/kinetic/backend.js]: Not destroying the socket seems to have same performances but seems less stable

Comment (AntoninCoulibaly) [inline lib/data/kinetic/backend.js]: So I investigated and it seems to come from StreamToPDU that using posix-read. So I guess the kineticlib will to have some changes. I think it could be better to have this changes in another PR.

File: lib/data/kinetic/backend.js (hunk 21 above)
Comment (lamphamsy) [inline lib/data/kinetic/backend.js]: We need it but it should be limited by `Number.MAX_VALUE`

**Comment by lamphamsy** [comment]
LGTM 👍

**Comment by vrancurel** [comment]
duplicate of #288
